R/calcSharesEU.R

```diff
@@ -0,0 +1,495 @@
+                rename("valueETP" = "value"),
+              by = c("region", "period", "enduse")) %>%
+    left_join(sharesChina %>%
+                rename("valueChina" = "value"),
+              by = c("region", "period", "enduse")) %>%
+    left_join(sharesEEI %>%
+                rename("valueEEI" = "value"),
+              by = c("region", "period", "enduse")) %>%
+    left_join(sharesWEO %>%
+                filter(.data$region != "CHN") %>%
+                rename("valueWEO" = "value"),
+              by = c("region", "period", "enduse")) %>%
+
+    # Find region-enduse pairs that have data in any non-ETP dataset
+    group_by(across(all_of(c("region", "enduse")))) %>%
+    mutate(
+      hasOtherData = any(
+        .data$period != 2014 & (
+          (!is.na(.data$value) & .data$value != 0) |             # Odyssee data
+            (!is.na(.data$valueChina) & .data$valueChina != 0) | # China data
+            (!is.na(.data$valueWEO) & .data$valueWEO != 0) |     # WEO data
+            (!is.na(.data$valueEEI) & .data$valueEEI != 0)       # EEI data
+        )
+      ),
+
+      # Remove ETP 2025 data for pairs that have other data or are WEO regions
+      valueETP = ifelse((.data$period == 2025 & .data$hasOtherData) |
+                          (.data$region %in% weoRegionsISO), NA, .data$valueETP),
+
+      # Count EEI data points for this region-enduse combination
+      eeiCount = sum(!is.na(.data$valueEEI) & .data$valueEEI != 0)
+    ) %>%
+    ungroup() %>%
+
+    # Apply priority rules with the WEO region condition
+    mutate(value = case_when(
+      !is.na(.data$value) & .data$value != 0                            ~ .data$value,      # 1. Odyssee
+      !is.na(.data$valueEEI)                                            ~ .data$valueEEI,   # 2. IEA EEI
+      !is.na(.data$valueChina)                                          ~ .data$valueChina, # 3. IEA China
+      (!is.na(.data$valueWEO) & .data$eeiCount < 2)                     ~ .data$valueWEO,   # 4. IEA WEO
+      (!is.na(.data$valueETP) & .data$eeiCount < 2)                     ~ .data$valueETP,   # 5. IEA ETP
+      TRUE                                                              ~ .data$value       # Fallback to Odyssee
+    )) %>%
+    select("region", "period", "enduse", "value")
+
+
+  # Extra-/Interpolate missing periods
+  sharesFilled <- shares %>%
+    .fillTimeseries(extGrowthFactors, periods = seq.int(periodBegin, endOfHistory, 1)) %>%
+    .normalize("enduse")
+
+
+  # Transform end-use "appliances" into "refrigerator" is specified
+  if (isTRUE(thermal)) {
+    sharesFilled <- sharesFilled %>%
+      toolAddThermal(regmappingEDGE, fridgeShare, feOnly = FALSE, "enduse") %>%
+      .normalize("enduse")
+  }
+
+
+  # Process final energy weights for regional aggregation
+  weights <- fe %>%
+    group_by(across(all_of(c("region", "period")))) %>%
+    reframe(value = sum(.data$value, na.rm = TRUE)) %>%
```

Comment:
I might be wrong but I don't think you still need the grouping afterwards. It should be faster if you ungroup in these cases.
```suggestion
    reframe(value = sum(.data$value, na.rm = TRUE)) %>%
    ungroup() %>%
```

Reply:
As I understand the `reframe` function, it automatically ungroups after execution, at least that has been my observation so far.